### ISSUE TITLE
Display the RSS memory in the boxplots of report.html and in the timeline.html instead of VMEM

### ISSUE THREAD

**Issue by phupe** [nextflow-io/nextflow#980]
## New feature

Follow-up to the discussion here https://groups.google.com/forum/#!topic/nextflow/qDzmI8YTy5A, we suggest to display the RSS memory in the boxplots of report.html and in the timeline.html instead of VMEM.

## Usage scenario 

As RSS memory corresponds to the memory held in the RAM, reporting this metrics really helps to calibrate the amount of memory resources to ask to the job scheduler when submitting the job on a cluster.

**Comment by pditommaso** [comment]
@phupe @ewels I've changed the memory report as discussed. The main figure now is the physical memory aka RAM or RSS. 

Please have look: 

<img width="1216" alt="screenshot 2019-01-30 at 14 37 17" src="https://user-images.githubusercontent.com/816968/51984911-dd6e4400-249c-11e9-897e-ab8ceeba0784.png">
<img width="1216" alt="screenshot 2019-01-30 at 14 37 19" src="https://user-images.githubusercontent.com/816968/51984925-e3642500-249c-11e9-82f4-650641746baf.png">


A couple things that remain to fix for which I would like to ask your help: 
1) the memory on the y-axis should be reported in base 2 instead of base 10 scale. That's controlled by the Plotly `tickformat` attribute, but I was not able to find a solution. 
2) hide the `% allocated` chart  when the data is missing (no `memory` is specified in the process) replacing it with a `Missing data` pane. 

Any suggestion?

**Comment by ewels** [comment]
1. I would think it's easier to pre-process the numbers in the javascript to MB before plotting. This should be pretty doable.
2. Yes, should definitely be doable. Need to play around to see what actually happens, but I guess that it will be a case of having some conditional javascript that fires when there's an empty data array / `undefined`/ `NaN` values etc. Typically I'd have a box with the warning (`<div class="alert alert-info" id="mem_warning" style="display:none;">No <code>memory</code> attributes set in this workflow</div>`) which is hidden on page load (`style="display:none;"`) and then shown if needed by jQuery (`$('#mem_warning').show()`).

**Comment by ewels** [comment]
Without actually getting in to the code with a test case, I think the code for part (2) is here:
https://github.com/nextflow-io/nextflow/blob/a9381f1ba88970b693a8ddf18685c4eaa4d36c00/modules/nextflow/src/main/resources/nextflow/trace/assets/ReportTemplate.js#L89-L93

Change to something like this:

```javascript
$('#pctmemplot_tablink').on('shown.bs.tab', function (e) { 
   if($('#pctmemplot').is(':empty')){
       if(mem_usage_data.length == 0){
          $('#pctmemplot').html('<div class="alert alert-info">No <code>memory</code> attributes set in this workflow</div>');
       } else {
          Plotly.newPlot('pctmemplot', mem_usage_data, { title: '% Requested Physical Memory Used', yaxis: {title: '% Memory', tickformat: '.1f', rangemode: 'tozero'} }); 
       }
   } 
 }); 
```
(ignore the `display:none` and jquery `show()` stuff)

**Comment by phupe** [comment]
Looks great! Thanks.

It seems that if you set  `type: 'log'` and `dtick` to `log10(2^10)` (= 3.0103) you may get what you want in order to have a log-scale in the y-axis with interval every 2^(n*10). See for example this plot in the following html file:

```

<head>
  <!-- Plotly.js -->
  <script src="https://cdn.plot.ly/plotly-latest.min.js"></script>
</head>

<body>
  
  <div id="myDiv" style="width: 480px; height: 400px;"><!-- Plotly chart will be drawn inside this DIV --></div>
  <script>
    var trace1 = {
    x: [.01, .1, 1, 10, 100], 
    y: [1, 1024, 1048576, 1073741824, 1099511628000], 
    type: 'scatter'
    };

    var data = [trace1];
    var layout = {
    xaxis: {
    type: 'log',
    autorange: true
    }, 

    yaxis: {
    type: 'log',
    dtick: 3.0103,
    }    
    };
    Plotly.plot('myDiv', data, layout, {showSendToCloud: true});
  </script>
</body>

```


But maybe the best option is to preprocess the data as suggested by @ewels .

Hope this helps.

**Comment by pditommaso** [comment]
Thanks for your feedback. Unfortunately the log scale trick does not seem to work. I'm trying to normalise the decimal numbers to binary scale using this javascript function: 

```javascript
function base2mem( value ) {
  var x = Math.floor(Math.log10(value) / Math.log10(1024) );
  if( x == 0 )
    value = value/1.024;
  else {
    for( i=0; i<x; i++ )
      value = value / 1.024;
  }
  return Math.round(value);
}
```

Then using it to pass the normalised value to Plotly as shown below: 

```
mem_raw_data.push({y: smry.mem, name: pname, type:'box', boxmean: true, boxpoints: false});
```

However Plotly refuse to render it(!) Any clue about what could be reason? I've noticed the original value has a `object` type, instead value returning the function is `number`. Why it should break plotly ?

**Comment by ewels** [comment]
Where is `base2mem` being called? I don't see it in the Plotly function call? If you're swapping an object for a value then plotly won't work, as you can't make a box plot with a single value and no categories... :)

**Comment by pditommaso** [comment]
Oops, i've pasted the wrong line before. I'm calling like this 

```
    mem_raw_data.push({y: base2mem(smry.mem), name: pname, type:'box', boxmean: true, boxpoints: false});
```

**Comment by pditommaso** [comment]
Ooook, I've understood the problem, `smry.mem` is an object with multiple values.

**Comment by ewels** [comment]
Yep 👍

**Comment by pditommaso** [comment]
🤦‍♂️ 🤦‍♂️

**Comment by pditommaso** [comment]
OK, fixed the y-scale for memory values.

**Comment by pditommaso** [comment]
Included in version `19.02.0-edge`.
